blocks/bricks/recurrent.py

```diff
@@ -630,3 +631,331 @@ def apply(self, *args, **kwargs):
+        usage of different docrations (wrappers) to be used.
+
+        Parameters
+        ----------
+        See docstring of the class for arguments appearing in
+        self.apply.sequences, self.apply.states, self.apply.contexts
+        All arguments values are of type :class:`~tensor.TensorVariable`.
+
+        In addition the `iterate`, `reverse`, `return_initial_states` or
+        any other argument defined in `recurrent_apply` wrapper.
+
+        Returns
+        -------
+        The outputs of all transitions as defined in `self.apply.outputs`
+        All return values are of type :class:`~tensor.TensorVariable`.
+
+        """
+        nargs = len(args)
+        assert nargs <= len(self.apply.sequences)
+        kwargs.update(zip(self.apply.sequences[:nargs], args))
+
+        if kwargs.get("reverse", False):
+            raise Exception("reverse=True not implemented.")
+
+        results = []
+        last_states = None
+        for level, transition in enumerate(self.transitions):
+            normal_inputs = self.normal_inputs(level)
+            layer_kwargs = dict()
+
+            if level == 0 or self.skip_connections:
+                for name in normal_inputs:
+                    layer_kwargs[name] = kwargs.get(self.suffix(name, level))
+            if "mask" in transition.apply.sequences:
+                layer_kwargs["mask"] = kwargs.get("mask")
+
+            for name in transition.apply.states:
+                layer_kwargs[name] = kwargs.get(self.suffix(name, level))
+
+            for name in transition.apply.contexts:
+                layer_kwargs[name] = kwargs.get(name)  # contexts has no suffix
+
+            if level > 0:
+                # add the forked states of the layer below
+                inputs = self.forks[level - 1].apply(last_states, as_list=True)
+                for name, input_ in zip(normal_inputs, inputs):
+                    if layer_kwargs.get(name):
+                        layer_kwargs[name] += input_
+                    else:
+                        layer_kwargs[name] = input_
+
+            # Handle all other arguments
+            # For example, if this method is called directly (from fast_apply)
+            # then these arguments can be the same arguments that recurrent
+            # expects to see such as: 'iterate', 'reverse',
+            # 'return_initial_states'
+            for k in set(kwargs.keys()) - self.transition_args:
+                layer_kwargs[k] = kwargs[k]
+
+            result = transition.apply(as_list=True, **layer_kwargs)
+            results.extend(result)
+
+            state_index = transition.apply.outputs.index(self.states_name)
+            last_states = result[state_index]
+            if kwargs.get('return_initial_states', False):
+                # Note that the following line reset the tag
+                last_states = last_states[1:]
+
+        return tuple(results)
+
+    @recurrent
+    def low_memory_apply(self, *args, **kwargs):
```

Comment:
I am worried that the way you provide the choice between the fast and the memory efficient one is rather quirky and that the `apply` method ends up having no docs at all. Alternatively one can go this way:

```
def do_apply(self, *args, **kwargs):
    pass

@recurrent
def low_memory_apply(self, *args, **kwargs):
    return self.do_apply(*args, **kwargs)

@application
def apply(self, *args, **kwargs):
    # all docs go here!
    if kwargs.get('low_memory', False):
        return self.low_memory_apply(*args, **kwargs)
    return self.do_apply(*args, **kwargs)
```

As a consequence the `fast` flag would be not required.